### PR TITLE
Do not allow to assign codes from an unavailable coupon

### PR DESCRIPTION
**Pull request by muhammad-ammar**
**Description:** Do not allow to assign codes from an unavailable coupon.

**Unavailable Coupon:** `coupon start datetime < current datetime < coupon end datetime`

**JIRA:** [ENT-2467](https://openedx.atlassian.net/browse/ENT-2467)

### CODE REVIEW

File: ecommerce/extensions/api/v2/tests/views/test_enterprise.py [hunk 1]
```diff
@@ -2383,6 +2385,27 @@ def test_coupon_codes_remind_all_with_invalid_code_filter(self):
         response = response.json()
         assert response == ['Invalid code_filter specified: invalid-filter']
 
+    @mock.patch('ecommerce.extensions.offer.utils.send_offer_assignment_email.delay', mock.Mock(return_value=None))
+    def test_expired_coupon_codes_assign(self):
+        """Test assigning codes from an expired coupon returns expected error reponse."""
+        start_datetime = self.data['start_datetime']
+        coupon_post_data = dict(self.data, end_datetime=start_datetime)
```
Comment (irfanuddinahmad) [inline ecommerce/extensions/api/v2/tests/views/test_enterprise.py:2392]: You could add more test cases with different combinations of start and end dates

File: ecommerce/coupons/tests/test_utils.py [hunk 2]
```diff
@@ -72,3 +81,44 @@ def test_fetch_course_catalog(self):
...
+            'start_datetime': datetime(2012, 11, 15, 1, 30, tzinfo=UTC),
+            'end_datetime': datetime(2012, 11, 15, 1, 30, tzinfo=UTC),
+            'timezone_now': datetime(2012, 11, 15, 1, 30, tzinfo=UTC),
+            'counpon_expired': True
+        },
+        {
+            'start_datetime': datetime(2012, 11, 15, 1, 30, tzinfo=UTC),
+            'end_datetime': datetime(2012, 11, 15, 1, 40, tzinfo=UTC),
+            'timezone_now': datetime(2012, 11, 15, 1, 35, tzinfo=UTC),
+            'counpon_expired': False
```
Comment (irfanuddinahmad) [inline ecommerce/coupons/tests/test_utils.py]: nit: counpon -> coupon

File: ecommerce/coupons/utils.py [hunk 3]
```diff
@@ -172,3 +173,14 @@ def is_voucher_applied(basket, voucher):
...
+
+def is_coupon_expired(coupon):
+    """
+    Returns True if `coupon` is expired, False otherwise.
+    """
+    voucher = coupon.attr.coupon_vouchers.vouchers.first()
+    start_datetime = voucher.start_datetime
+    end_datetime = voucher.end_datetime
+    current_datetime = timezone.now()
+    return not start_datetime < current_datetime < end_datetime
```
Comment (irfanuddinahmad) [inline ecommerce/coupons/utils.py]: This function tags future dated coupons as expired ... do we need to handle this here

File: ecommerce/coupons/utils.py [hunk 4]
```diff
@@ -172,3 +173,14 @@ def is_voucher_applied(basket, voucher):
         if discount['voucher'] and discount['voucher'] == voucher:
             return True
     return False
+
+
+def is_coupon_active(coupon):
```
Comment (irfanuddinahmad) [inline ecommerce/coupons/utils.py]: is_coupon_active -> is_coupon_temporally_active

Comment (muhammad-ammar) [reply ecommerce/coupons/utils.py]: I would prefer to use `is_coupon_active`

**Review by waheedahmed — APPROVED** [review]

File: ecommerce/coupons/utils.py (hunk 4 above)
Comment (irfanuddinahmad) [reply ecommerce/coupons/utils.py]: There is some work going on regarding active statuses as part of https://openedx.atlassian.net/browse/ENT-2445
The temporal prefix will reduce chances of confusion with active statuses that are not based on start/end dates.
Alternately, we could use 'is_coupon_date_valid' or something similar ...

**Review by irfanuddinahmad — CHANGES_REQUESTED** [review]

**Comment by muhammad-ammar** [comment]
@irfanuddinahmad I have addressed your feedback.

**Review by irfanuddinahmad — APPROVED** [review]
LGTM!

**Pull request merged**